### PR TITLE
fixes #2420 - extract strings for i18n from JavaScript, various i18n fixes

### PR DESCRIPTION
**Pull request by domcleal**
cc @lzap, @abenari for review

The main bits that are missing are:
- will_paginate (uses Rails i18n, was working on this separately)
- jquery datatable (jquery has its own i18n system that needs looking into)
- any other asset gems, e.g. spice-html5

I wanted to do this initially as a separate i18n resource/domain, but gettext_i18n_rails_js didn't support that, so this will have to do for now.  The resulting JSON files aren't that big anyway.

### CODE REVIEW

File: app/assets/javascripts/charts.js [hunk 1]
```diff
@@ -231,8 +231,8 @@ function get_pie_chart(div, url) {
   if($("#"+div).length == 0)
   {
     $('body').append('<div id="' + div + '" class="modal fade"></div>');
-    $("#"+div).append('<div class="modal-header"><a href="#" class="close" data-dismiss="modal">&times;</a><h3>Fact Chart</h3></div>')
-        .append('<div id="' + div + '-body" class="fact_chart modal-body">Loading ...</div>');
+    $("#"+div).append('<div class="modal-header"><a href="#" class="close" data-dismiss="modal">&times;</a><h3>' + _('Fact Chart') + '</h3></div>')
+        .append('<div id="' + div + '-body" class="fact_chart modal-body">' + _('Loading ...') + '</div>');
```
Comment (lzap) [inline app/assets/javascripts/charts.js]: Minor: "Loading" + " ..."

File: app/helpers/audits_helper.rb [hunk 2]
```diff
@@ -4,10 +4,10 @@ module AuditsHelper
 
   # lookup the Model representing the numerical id and return its label
   def id_to_label name, change
-    return "N/A" if change.nil?
+    return _("N/A") if change.nil?
     case name
       when "ancestry"
-        change.blank? ? "" : change.split('/').map { |i| Hostgroup.find(i).name rescue "NA" }.join('/')
+        change.blank? ? "" : change.split('/').map { |i| Hostgroup.find(i).name rescue _("NA") }.join('/')
```
Comment (lzap) [inline app/helpers/audits_helper.rb:10]: Check: NA or N/A?

**Comment by lzap** [comment]
Looking good thanks for those fixes of Ruby codebase. :shipit:

**Comment by domcleal** [comment]
I've opened a PR against gettext_i18n_rails_js to fix the 1.8 compat: https://github.com/nubis/gettext_i18n_rails_js/pull/17

File: app/helpers/audits_helper.rb (hunk 2 above)
Comment (domcleal) [inline app/helpers/audits_helper.rb:10]: I wondered about this and decided it was deliberate, as it's using forward slashes to separate up the hostgroup hierarchy, so presumably the original author thought the slashes in "N/A" would get confusing.

File: app/assets/javascripts/charts.js (hunk 1 above)
Comment (domcleal) [inline app/assets/javascripts/charts.js]: ta, fixed!

**Comment by domcleal** [comment]
I've rebased the branch and it now depends on the gettext_i18n_rails_js release with Ruby 1.8 support.  (cc @ohadlevy for merge)

File: app/assets/javascripts/config_template.js [hunk 3]
```diff
@@ -52,12 +52,12 @@ $(function() {
...
-      Default: null, // Null = use "default" keymapping
-      Vim: vim,
-      Emacs: emacs};
+    var keybindings = [
+      null, // Null = use "default" keymapping
+      vim,
+      emacs];
 
-    $editor.setKeyboardHandler(keybindings[$("#keybinding").val()]);
+    $editor.setKeyboardHandler(keybindings[$("#keybinding")[0].selectedIndex]);
```
Comment (ohadlevy) [inline app/assets/javascripts/config_template.js:60]: why this change? /cc @abenari

File: app/assets/javascripts/host_edit.js [hunk 4]
```diff
@@ -434,7 +434,7 @@ function reload_puppetclass_params(){
 function load_with_placeholder(target, url, data){
   if(url==undefined) return;
   var placeholder = $('<tr id="' + target + '_loading" >'+
-            '<td colspan="4"><p><img src="/assets/spinner.gif" alt="Wait" /> Loading parameters...</p></td></tr>');
+            '<td colspan="4"><p><img src="/assets/spinner.gif" alt="' + _('Wait') + '" /> ' + _('Loading parameters...') + '</p></td></tr>');
```
Comment (ohadlevy) [inline app/assets/javascripts/host_edit.js:437]: I'm wondering if this should not be a helper

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -243,18 +243,29 @@ def render_403
...
+    unless hash[:success_msg]
+      hash[:success_msg] = case action_name
+                           when "create"
+                             _("Successfully created %s.") % hash[:object_name]
+                           when "update"
+                             _("Successfully updated %s.") % hash[:object_name]
+                           when "destroy"
+                             _("Successfully deleted %s.") % hash[:object_name]
+                           else
+                             raise
```
Comment (ohadlevy) [inline app/controllers/application_controller.rb]: raise what? shouldnt we use Foreman::Exception.new ?

Comment (ohadlevy) [inline app/controllers/application_controller.rb]: also, imho this is way to verbose, why not:

``` ruby
action = action_name.pluralize.sub(/es$/,"ed").sub(/ys$/, "yed")
hash[:success_msg = _("Successfully #{action} %s.") % action, hash[:object_name]
```

File: app/assets/javascripts/config_template.js (hunk 3 above)
Comment (domcleal) [inline app/assets/javascripts/config_template.js:60]: I extracted the strings in the view used in the dropdown list (so "Default" could be translated), which means looking up the keymapping using a hash won't work with a translated word.  This changes it to look up by index instead.

File: app/assets/javascripts/host_edit.js (hunk 4 above)
Comment (domcleal) [inline app/assets/javascripts/host_edit.js:437]: I wasn't trying to refactor code too :)

File: app/helpers/audits_helper.rb [hunk 6]
```diff
@@ -83,7 +83,7 @@ def audit_user audit
   end
 
   def audit_time audit
-    content_tag :span, time_ago_in_words(audit.created_at) + " ago",
+    content_tag :span, _("%s ago") % time_ago_in_words(audit.created_at),
```
Comment (ohadlevy) [inline app/helpers/audits_helper.rb:86]: not sure if this would work, afair the output is something like
'5 days'

so you'll end up having "5 days %s"

File: app/controllers/application_controller.rb (hunk 5 above)
Comment (domcleal) [inline app/controllers/application_controller.rb]: I will fix the exception.

We can't simplify it to that as the full strings won't get extracted and made available to translators.  They'll only see "Successfully %s %s" and the action words won't get translated.

The closest you could do I think would be to extract "created", "updated", "deleted" separately to "Successfully %s %s", but then that gets confusing for translators and could cause issues in some languages.

File: app/helpers/audits_helper.rb (hunk 6 above)
Comment (domcleal) [inline app/helpers/audits_helper.rb:86]: should end up as "5 days ago"?

Comment (ohadlevy) [inline app/helpers/audits_helper.rb:86]: no, what i mean is that 'days' wont be extracted?

Comment (domcleal) [inline app/helpers/audits_helper.rb:86]: Ah no, I think that gets handled entirely by Rails' locale impl.  On the audit page I now see: "Xetwa 17 Stunden agoX" when switching my locale to German.

**Comment by domcleal** [comment]
@ohadlevy rebased, exception fixed

**Comment by ohadlevy** [comment]
merged - thanks
